PhysicsTools/PatAlgos/plugins/DisplacedMuonFilterProducer.cc

```diff
@@ -0,0 +1,305 @@
+    timeMapDTToken_ = consumes<reco::MuonTimeExtraMap>(edm::InputTag(srcMuons_.label(), "dt"));
+    timeMapCSCToken_ = consumes<reco::MuonTimeExtraMap>(edm::InputTag(srcMuons_.label(), "csc"));
+
+    produces<reco::MuonTimeExtraMap>("combined");
+    produces<reco::MuonTimeExtraMap>("dt");
+    produces<reco::MuonTimeExtraMap>("csc");
+  }
+
+  if (fillDetectorBasedIsolation_) {
+    theTrackDepositName = iConfig.getParameter<edm::InputTag>("TrackIsoDeposits");
+    theTrackDepositToken_ = consumes<reco::IsoDepositMap>(theTrackDepositName);
+
+    theJetDepositName = iConfig.getParameter<edm::InputTag>("JetIsoDeposits");
+    theJetDepositToken_ = consumes<reco::IsoDepositMap>(theJetDepositName);
+
+    theEcalDepositName = iConfig.getParameter<edm::InputTag>("EcalIsoDeposits");
+    theEcalDepositToken_ = consumes<reco::IsoDepositMap>(theEcalDepositName);
+
+    theHcalDepositName = iConfig.getParameter<edm::InputTag>("HcalIsoDeposits");
+    theHcalDepositToken_ = consumes<reco::IsoDepositMap>(theHcalDepositName);
+
+    theHoDepositName = iConfig.getParameter<edm::InputTag>("HoIsoDeposits");
+    theHoDepositToken_ = consumes<reco::IsoDepositMap>(theHoDepositName);
+
+    produces<reco::IsoDepositMap>("tracker");
+
+    produces<reco::IsoDepositMap>("ecal");
+
+    produces<reco::IsoDepositMap>("hcal");
+
+    produces<reco::IsoDepositMap>("ho");
+
+    produces<reco::IsoDepositMap>("jets");
+  }
+}
+
+pat::DisplacedMuonFilterProducer::~DisplacedMuonFilterProducer() {}
+
+// Filter muons
+
+void pat::DisplacedMuonFilterProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  auto output = std::make_unique<reco::MuonCollection>();
+
+  // muon collections
+  edm::Handle<reco::MuonCollection> srcMuons;
+  iEvent.getByToken(srcMuonToken_, srcMuons);
+
+  int nMuons = srcMuons->size();
+
+  // filter the muons
+  std::vector<bool> filteredmuons(nMuons, true);
+  int oMuons = nMuons;
+
+  for (unsigned int i = 0; i < srcMuons->size(); i++) {
+    const reco::Muon& muon(srcMuons->at(i));
+
+    if (muon.isStandAloneMuon()) {
+      if (muon.innerTrack().isNonnull()) {
+        // Discard STA + GL/TR muons that are below pt threshold
+        if (muon.innerTrack()->pt() < minPtTK_ && muon.standAloneMuon()->pt() < minPtSTA_) {
+          filteredmuons[i] = false;
+          oMuons = oMuons - 1;
+          continue;
+        }
+      } else {
+        // Discard STA-only muons with less than minMatches_ segments and below pt threshold
+        if (!muon.isMatchesValid() || muon.numberOfMatches() < minMatches_ || muon.standAloneMuon()->pt() < minPtSTA_) {
+          filteredmuons[i] = false;
+          oMuons = oMuons - 1;
+          continue;
+        }
+      }
+    } else {
+      if (muon.innerTrack().isNonnull()) {
+        if (muon.innerTrack()->pt() < minPtTK_ || !muon.isTrackerMuon()) {
+          filteredmuons[i] = false;
+          oMuons = oMuons - 1;
+          continue;
+        }
+      } else {  // Should never happen
```

Comment:
Maybe it would be good to add an `edm::LogWarning`

Reply:
Added